aten/src/ATen/native/cuda/Distributions.cu

```diff
@@ -441,11 +443,77 @@ void uniform_kernel_cuda(TensorIterator& iter, double from_, double to_, Generat
    });
 }
 
+void random_kernel_cuda(TensorIterator& iter, uint64_t range, int64_t base, Generator* gen_) {
+  auto gen = check_generator<CUDAGenerator>(gen_, &globalContext().defaultGenerator(kCUDA));
+  AT_DISPATCH_ALL_TYPES_AND(at::ScalarType::Half, iter.dtype(), "random_cuda", [&] {
+    using accscalar_t = at::acc_type<scalar_t, true>;
+    if (std::is_same<scalar_t, double>::value || std::is_same<scalar_t, int64_t>::value) {
+      // define lambda to mod with range and add base
+      auto random_func = [range, base] __device__ (uint64_t rand) {
+        return static_cast<scalar_t>(rand % range + base);
+      };
+      distribution_nullary_kernel<scalar_t, accscalar_t, curand4_engine_calls/2>(iter,
+        gen,
+        [] __device__ (curandStatePhilox4_32_10_t* state) -> ulonglong2 {
+          ulonglong2 ret;
+          uint4 rand = curand4(state);
+          ret.x = (static_cast<uint64_t>(rand.x) << 32) | rand.y;
+          ret.y = (static_cast<uint64_t>(rand.z) << 32) | rand.w;
+          return ret;
+        },
+        random_func);
+    } else {
+      auto random_func = [range, base] __device__ (uint32_t rand) {
+        return static_cast<scalar_t>(rand % static_cast<uint32_t>(range) + static_cast<int32_t>(base));
+      };
+      distribution_nullary_kernel<scalar_t, accscalar_t, curand4_engine_calls>(iter,
+        gen,
+        [] __device__ (curandStatePhilox4_32_10_t* state) { return curand4(state); },
+        random_func);
+    }
+   });
+}
+
 Tensor& uniform_cuda_(Tensor& self, double from, double to, Generator* gen) {
   auto iter = TensorIterator::nullary_op(self);
   uniform_kernel_cuda(*iter, from, to, gen);
   return self;
 }
 
+Tensor& random_cuda_(Tensor& self, Generator* gen) {
+  auto iter = TensorIterator::nullary_op(self);
+  uint64_t range;
+  auto iter_scalar_type = iter->dtype();
+  if (isFloatingType(iter_scalar_type)) {
+    AT_DISPATCH_FLOATING_TYPES_AND_HALF(iter_scalar_type, "random_cuda_range_calc", [&] {
+      range = static_cast<uint64_t>((1ULL << std::numeric_limits<scalar_t>::digits) + 1);
+    });
+  } else {
+    AT_DISPATCH_INTEGRAL_TYPES(iter_scalar_type, "random_cuda_range_calc", [&] {
+      range = static_cast<uint64_t>(std::numeric_limits<scalar_t>::max()) + 1;
+    });
+  }
+  random_kernel_cuda(*iter, range, 0, gen);
+  return self;
+}
+
+Tensor& clamped_random_cuda_(Tensor& self, int64_t from, int64_t to, Generator* gen) {
+  AT_CHECK(from < to, "random_ expects 'from' to be less than 'to', but got from=", from, " >= to=", to);
+  auto iter = TensorIterator::nullary_op(self);
+  AT_DISPATCH_ALL_TYPES_AND(at::ScalarType::Half, iter->dtype(), "range_check", [&] {
```

Comment:
Man, this dispatch /just/ to do error checking makes my eyes bleed, but I can't think of a convenient way to do this otherwise, so fine.

Reply:
I removed this range_check btw because of bool tensors and that we don't seem to guarantee this condition (didn't had this check before and the following says "usually"):
```
random_(from=0, to=None, *, generator=None) → Tensor
Fills self tensor with numbers sampled from the discrete uniform distribution over [from, to - 1]. If not specified, the values are usually only bounded by self tensor’s data type. However, for floating point types, if unspecified, range will be [0, 2^mantissa] to ensure that every value is representable. For example, torch.tensor(1, dtype=torch.double).random_() will be uniform in [0, 2^53].
```

We should have this discussion in https://github.com/pytorch/pytorch/issues/16944. As pointed by @ngimel, following is the behavior in numpy:
```In [3]: np.random.randint(0,500, size=100, dtype=np.int8)                                                                                            
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-587d8ebfc2e9> in <module>
----> 1 np.random.randint(0,500, size=100, dtype=np.int8)

mtrand.pyx in mtrand.RandomState.randint()

ValueError: high is out of bounds for int8
```
Same for bool
```In [6]: np.random.randint(0,3, size=100, dtype=np.bool)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-6-8c6e78ec7bf7> in <module>
----> 1 np.random.randint(0,3, size=100, dtype=np.bool)

mtrand.pyx in mtrand.RandomState.randint()

ValueError: high is out of bounds for bool
```

Reply:
I don't think this discussion belongs in #16944. #16944 is about valid arguments to random_, that are not working currently, and the reasons for them not working are clear.  
Things we need to decide (and possibly change behavior) are what happens when `random_` is called with no arguments  - right now behavior is inconsistent. The docs say `the values are usually only bounded by self tensor’s data type` but try generating negative numbers for `int` tensor. My suggestion here is to use full range, but #16944 will have to be fixed for this. I'm also not against documenting and preserving current behavior. Also, what happens when `random_` arguments fall outside numeric range of source tensor - my suggestion is to follow numpy and throw an error.